libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/offboard/MapboxOffboardRouter.kt

```diff
@@ -111,29 +134,79 @@ class MapboxOffboardRouter(
         route: DirectionsRoute,
         legIndex: Int,
         callback: RouteRefreshCallback
-    ) {
-        try {
-            val refreshBuilder = RouteBuilderProvider.getRefreshBuilder()
-                .accessToken(accessToken)
-                .requestId(route.routeOptions()?.requestUuid())
-                .routeIndex(route.routeIndex()?.toIntOrNull() ?: 0)
-                .legIndex(legIndex)
-                .interceptor {
-                    val httpUrl = (it.request() as Request).url
-                    val skuUrl = urlSkuTokenProvider.obtainUrlWithSkuToken(httpUrl.toUrl())
-                    it.proceed(it.request().newBuilder().url(skuUrl).build())
+    ): Long {
+        val mapboxDirectionsRefresh = RouteBuilderProvider.getRefreshBuilder()
+            .accessToken(accessToken)
+            .requestId(route.routeOptions()?.requestUuid())
+            .routeIndex(route.routeIndex()?.toIntOrNull() ?: 0)
+            .legIndex(legIndex)
+            .interceptor {
+                val httpUrl = (it.request() as Request).url
+                val skuUrl = urlSkuTokenProvider.obtainUrlWithSkuToken(httpUrl.toUrl())
+                it.proceed(it.request().newBuilder().url(skuUrl).build())
+            }
+            .build()
+        val requestId = refreshRequests.put(mapboxDirectionsRefresh)
+
+        mapboxDirectionsRefresh.enqueueCall(object : Callback<DirectionsRefreshResponse> {
+            override fun onResponse(
+                call: Call<DirectionsRefreshResponse>,
+                response: Response<DirectionsRefreshResponse>
+            ) {
+                refreshRequests.remove(requestId)
+                if (response.isSuccessful) {
+                    val routeAnnotations = response.body()?.route()
+                    var errorThrowable: Throwable? = null
+                    val refreshedDirectionsRoute = try {
+                        RouteRefreshCallbackMapper.mapToDirectionsRoute(route, routeAnnotations)
+                    } catch (t: Throwable) {
+                        errorThrowable = t
+                        null
+                    }
+                    if (refreshedDirectionsRoute != null) {
+                        callback.onRefresh(refreshedDirectionsRoute)
+                    } else {
+                        callback.onError(
+                            RouteRefreshError(
+                                message = "Failed to read refresh response",
+                                throwable = errorThrowable ?: Exception(
+                                    "Message=[${response.message()}]; " +
+                                        "errorBody = [${response.errorBody()}];" +
```

Comment:
Should we extract this into a `constant` (`Message=[` and `"errorBody = [`)? This is duplicated down here.